src/uct/ib/rc/accel/rc_mlx5_ep.c

```diff
@@ -228,8 +240,113 @@ ucs_status_t uct_rc_mlx5_ep_get_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov, size
+    return sizeof(data->hdr) + data->super.length;
+}
+
+/* DM memory should be written by 8 bytes (int64) to eliminate
+ * processor cache issues. to make this used uct_rc_mlx5_dm_copy_data_t
+ * structure where first hdr_len bytes are filled by message header
+ * and tail is filled by head of message */
+static void UCS_F_ALWAYS_INLINE
+uct_rc_mlx5_ep_copy_to_dm(uct_rc_mlx5_dm_copy_data_t *cache, size_t hdr_len,
+                          const void *payload, size_t length, void *dm)
+{
+    size_t len          = sizeof(*cache) - hdr_len;
+    const uint64_t *src = payload + len;
+    uint64_t *dst       = dm;
+
+    ucs_assert(sizeof(*cache) >= hdr_len);
+
+    /* copy head of payload to tail of cache */
+    memcpy(cache->in + hdr_len, payload, ucs_min(len, length));
+
+    UCS_STATIC_ASSERT(sizeof(*cache) == sizeof(cache->out));
+    UCS_STATIC_ASSERT(sizeof(cache->in) == sizeof(cache->out));
+
+    /* atomically by 8 bytes copy data to DM */
+    *(dst++) = cache->out[0];
+    *(dst++) = cache->out[1];
+
+    for (; len < length; len += sizeof(*dst)) {
+        *(dst++) = *(src++);
+    }
+}
+
+ucs_status_t
+uct_rc_mlx5_ep_am_short_dm(uct_ep_h tl_ep, uint8_t id, uint64_t hdr,
+                           const void *payload, unsigned length)
+{
+    uct_rc_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_rc_mlx5_iface_t);
+    uct_rc_iface_t *rc_iface   = &iface->super;
+    uct_rc_mlx5_ep_t *ep       = ucs_derived_of(tl_ep, uct_rc_mlx5_ep_t);
+    uct_rc_iface_send_desc_t *desc;
+    size_t len;
+    uct_rc_mlx5_dm_pack_data_t pack_data;
+    uct_rc_mlx5_dm_copy_data_t cache;
+
+    if (sizeof(uct_rc_am_short_hdr_t) + length < UCT_IB_MLX5_AM_MAX_SHORT(0)) {
+        return uct_rc_mlx5_ep_am_short_inline(tl_ep, id, hdr, payload, length);
+    } else {
+        UCT_CHECK_LENGTH(length + sizeof(uct_rc_hdr_t), 0,
+                         iface->mlx5_common.dm.seg_len, "am_short");
+        UCT_CHECK_AM_ID(id);
+        UCT_RC_CHECK_RES(rc_iface, &ep->super);
+        UCT_RC_CHECK_FC(rc_iface, &ep->super, id);
+
+        desc = ucs_mpool_get_inline(&iface->mlx5_common.dm.dm->mp);
+        if (ucs_unlikely((desc) == NULL)) {
+            /* in case if no resources available - fallback to bcopy */
+            pack_data.hdr           = hdr;
+            pack_data.super.length  = length;
+            pack_data.super.payload = payload;
+
+            UCT_RC_IFACE_GET_TX_AM_BCOPY_DESC(rc_iface, &rc_iface->tx.mp, desc, id,
+                                              uct_rc_mlx5_ep_am_short_dm_pack,
+                                              &pack_data, &len);
+
+            uct_rc_mlx5_txqp_bcopy_post(rc_iface, &ep->super.txqp, &ep->tx.wq,
+                                        MLX5_OPCODE_SEND, sizeof(uct_rc_hdr_t) + len,
+                                        0, 0, MLX5_WQE_CTRL_SOLICITED, 0, desc, desc + 1);
+            UCT_TL_EP_STAT_OP(&ep->super.super, AM, BCOPY, len);
```

Comment:
can use goto here to avoid code duplication

Reply:
refactored